### PR TITLE
Refactor List diffing/change records, add diffing interface

### PR DESCRIPTION
**Pull request by matanlurey**
This hopefully can be used to share code between more and more libraries/frameworks that need to determine if a List changed since last checked. Angular Dart has their own implementation here:

https://github.com/dart-lang/angular2/blob/master/lib/src/core/change_detection/differs/default_iterable_differ.dart

This first implementation just establishes a `ListDiffer` that is a copy of the original work - we can improve/speed it up in future PRs.

(This goes without saying, but I'll squash before merging)

## 0.15.0

* Added the `Differ` interface, as well as `EqualityDiffer`
* Refactored list diffing into a `ListDiffer`
* Added concept of `ChangeRecord.ANY` and `ChangeRecord.NONE`
    * Low-GC ways to expression "something/nothing" changed
* Refactored `ListChangeRecord`
    * Added named constructors for common use cases
    * Added equality and hashCode checks
    * Added `ListChangeRecord.apply` to apply a change record
* Added missing `@override` annotations to satisfy `annotate_overrides`

### CODE REVIEW

File: lib/src/differs.dart [hunk 1]
```diff
@@ -0,0 +1,31 @@
+library observable.src.differs;
```
Comment (jmesserly) [inline lib/src/differs.dart:5]: copyright header? also FWIW I think library tags are optional now

**Review by jmesserly — APPROVED** [review]
LGTM!

File: lib/src/differs.dart (hunk 1 above)
Comment (matanlurey) [reply lib/src/differs.dart:5]: Not yet if you are using `part` 👎 

Added the header though, thanks. Wished the presubmit checked that!

Comment (jmesserly) [reply lib/src/differs.dart:5]: oh, right! booooo :)

**Pull request merged**